### PR TITLE
Workbox v4 updates

### PR DESCRIPTION
**Pull request by jeffposnick**
*Do not merge, pending the final Workbox v4 release.*

R: @philipwalton 

This contains the majority of the changes to the existing guides and recipes for Workbox to account for the syntax changes in v4.

This does not include:
- any new guides or recipes around `workbox-window`
- a Workbox v3 to v4 migration guide

[WF_IGNORE:TEMPLATE_TAGS,TYPOS]

### CODE REVIEW

File: src/content/en/tools/workbox/modules/workbox-core.md [hunk 1]
```diff
@@ -105,3 +66,23 @@ The main use case for the prefix and suffix is that if you use Workbox for
...
 custom prefix for each module will prevent the caches from conflicting
 with each other.
+
+## Skip Waiting and Clients Claim
+
+Some developers want to be able to publish a new service worker and have it
+update and control a web page as soon as possible, skipping the default
+[service worker lifecycle](/web/fundamentals/primers/service-workers/lifecycle).
+
+If you find yourself wanting this behavior, `workbox-sw` provides some helper
```
Comment (philipwalton) [inline src/content/en/tools/workbox/modules/workbox-core.md]: This is not in ` workbox-core` not `workbox-sw`.

**Comment by jeffposnick** [comment]
I think we can share the `wb-v4-updates` branch, and you could push your changes to it as well. We can then merge everything in one go in this PR.

(Just be warned that there are some false-positives in the pre-commit script due to how we use partial templates, and you'll need to use `git push --no-verify` to get additional commits pushed to this branch.)

**Comment by philipwalton** [comment]
@jeffposnick add a few things to the PR:

- Updated the "Advanced Recipes" section to include `workbox-window` examples.
- Added "Using Bundlers (webpack/Rollup) with Workbox" to the guides section and TOC.
- Added "Workbox Window Usage Guide" to the modules section, TOC, and index page.

File: src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+// Instead, use the constructor directly:
+// (Note that the class name is Uppercase.)
+const networkFirstStrategy = new workbox.strategies.NetworkFirst({...});
+```
+
+While the factory method syntax will continue working in v4, using it will log a
+warning, and we encourage developers to migrate in advance of removing support
+in the future v5 release.
+
+### Workbox-background-sync
```
Comment (philipwalton) [inline src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md]: Lowercase "Workbox" here.

File: src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+### workbox-window
+
+The new `workbox-window` module simplifies the process of service worker
+registration and detecting updates, and provides a standard means of
+communication between code running in the service worker and code running in
+your web app's `window` context.
+
+While using `workbox-window` is optional, we hope that developers will find it
+useful, and consider migrating some of their handwritten logic to use it when
+appropriate. You can learn more about using workbox-window in the [module's guide](/web/tools/workbox/modules/workbox-window).
```
Comment (philipwalton) [inline src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md]: s/workbox-window/`workbox-window`/

File: src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+- `pushRequest()`
+- `popRequest()`
+- `shiftRequest()`
+- `unshiftRequest()`
+
+In v3, the `Queue` class also accepted several callbacks that allowed you to
+observe when requests were being replayed (`requestWillEnqueue`,
+`requestWillReplay`, `queueDidReplay`), but most developers found that, in
+addition to observation, they wanted control over how the queue was replayed,
+including the ability to dynamically modify, re-order, or event cancel
```
Comment (philipwalton) [inline src/content/en/tools/workbox/guides/migrations/migrate-from-v3.md]: s/even/event/

**Pull request merged**